src/utils/join.js

```diff
@@ -1,7 +1,11 @@
-// For some reason path.posix.join is undefined in webpack
-// Also, this is just much smaller
-import { normalizePath } from './normalizePath.js'
+const { normalize } = require('path').posix || require('path')
 
 export function join(...parts) {
-  return normalizePath(parts.map(normalizePath).join('/'))
+  let normalizedPath = normalize(parts.map(normalize).join('/'))
```

Comment:
I don't think it's necessary to normalize each part passed in as normalize is already going to do that on the final result...so it's redundant.

Reply:
I agree with this assumption as well at first.
But it fails the tests (both nominal cases and special cases)

Reply:
I'd be interested to know why. That seems really odd to me. Unless the `.join('/')` is too naive and causing problems.

btw, in Node.js, path.join already calls normalize, so perhaps we could leverage that as well.

(sometimes the abstractions in isomorphic-git just seem way overdone).

Reply:
I think the difference with node `path.join` algorithm is that `isomorphic-git` "normalize" every parameters before joining them altogether and normalize the result, whereas node `path.join` normalize after joining every parts.

Pass all the tests (I'll commit that):
```javascript
let normalizedPath = path.join(...parts.map(path.normalize))
```

Does not pass the **when "internal join" generates paths differently from "path.join"** tests:
```javascript
let normalizedPath = path.join(...parts)
```

Reply:
I don't understand why isomorphic-git would rely on logic that normalizes each segment first. That just strikes me as odd. Maybe there is a good reason. I'd like to know which test case in particular is failing so we can understand why the behavior is different, and why it is relying on that behavior.

Reply:
Me too, I don't understand why `isomorphic-git` need this extra special cases... I don't see the use cases.

All the tests case **when "internal join" generates paths differently from "path.join"** using this implementation:
```javascript
let normalizedPath = path.join(...parts)
```
![Screenshot 2024-01-27 at 09 53 19](https://github.com/isomorphic-git/isomorphic-git/assets/522422/28487751-263e-4d7b-9501-8780e69492fc)
![Screenshot 2024-01-27 at 09 59 02](https://github.com/isomorphic-git/isomorphic-git/assets/522422/4cd46a8f-9b14-4e18-82bc-28fc9cd22559)
![Screenshot 2024-01-27 at 09 59 11](https://github.com/isomorphic-git/isomorphic-git/assets/522422/c56ef715-b5b4-4a3f-8a26-1ff9589cbea6)
![Screenshot 2024-01-27 at 09 59 20](https://github.com/isomorphic-git/isomorphic-git/assets/522422/13d2c2b8-6034-4296-a585-7bd1c7f691aa)


Those "special" test case are a bit surprising because they fail mostly by adding a `/` in front or in the end of the path, the front `/` is **not an issue** in the nominal cases and the trailing `/` could easily be trimmed.
The first two cases dealing with `./` can be treated the same as removing trailing `/`.
If those special cases can be avoided then we could write it this way:
```javascript
let normalizedPath = path.join(...parts)

  if (normalizedPath.endsWith('/')) {
    normalizedPath = normalizedPath.slice(0, -1)
  }

  return normalizedPath
```
And if trailing `/` is not important then we could use `path.join` directly imho

Reply:
If this is just to satisfy unit tests, then I'd say the tests are overzealous. From the comment at the top of normalizePath.js, it seems to me like the intention was to replace path.posix. If that's the case, it's path.posix to which we should be aligning, NOT this custom implementation.

Reply:
I totally agree on the principle, I did not want to force this kind of design as I assumed it was there for a good reason even if I don't know it.

Anyway, I just tried using `path.join` with my application and it works.
So I'll update this PR so the code uses the `src/utils/path.js` implementation

Reply:
Using the actual `src/utils/path.js` is not treeshakeable.
It is detected by `agadoo`.
It was not detected before because treeshakeability test is done after `rollup`, and, as `src/utils/path.js` was not used it was not included in `index.js` then it was not a problem for agadoo.
Now it is a problem because we use it instead of join...

I tried
- to make the export of path work, without success ❌
- upgrading `rollup`, without success ❌
- upgrading `agadoo`, without success ❌
- using another lib that behave always the same instead of path (`upath`), success ✅

Reply:
> I assumed it was there for a good reason even if I don't know it.

I looked at the history and there really doesn't seem to be a good reason other than to avoid double slash, which all these implementation options do.